File: packages/contracts/src/models/tasks.ts
import { z } from "zod";
import { taskIdSchema } from "./ids.js";

export const taskStateSchema = z.enum([
  "queued",
  "running",
  "completed",
  "failed",
  "cancelled",
]);

export const taskKindSchema = z.enum([
  "scan",
  "ingest",
  "normalize",
  "thumbnail",
  "ffmpeg",
  "subtitle",
]);

export const taskProgressSchema = z.object({
  taskId: taskIdSchema,
  taskType: taskKindSchema,
  state: taskStateSchema,
  current: z.number().int().nonnegative(),
  total: z.number().int().nonnegative().optional(),
  message: z.string().optional(),
  errorCode: z.string().optional(),
});

export type TaskState = z.infer<typeof taskStateSchema>;
export type TaskKind = z.infer<typeof taskKindSchema>;
export type TaskProgress = z.infer<typeof taskProgressSchema>;

File: crates/shared-model/src/tasks.rs
use crate::TaskId;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TaskState {
    Queued,
    Running,
    Completed,
    Failed,
    Cancelled,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TaskKind {
    Scan,
    Ingest,
    Normalize,
    Thumbnail,
    Ffmpeg,
    Subtitle,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TaskProgress {
    pub task_id: TaskId,
    pub task_type: TaskKind,
    pub state: TaskState,
    pub current: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub total: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_code: Option<String>,
}
